Fix the paper list.

codeviking opened the pull request:
A long time ago in a galaxy far, far away...I added a list of papers
that were interesting to look at in ScholarPhi.

This PR fixes that list, in hopes that it might help with things
like https://github.com/allenai/scholar/issues/25368.

The fix required revising a database query and removing a whitelist,
which is probably ok to remove since (I think) we're beyond whatever
demo window it was added for.

mjlangan commented on ui/src/papers/index.tsx:
```diff
@@ -162,17 +163,7 @@ const PaperList = () => {
-                      "1404.6635",
-                    ].indexOf(paper.arxivId) !== -1
-                  );
+                  return paper.arxivId !== undefined;
```
```suggestion
                  return !!paper.arxivId;
```
If we just want to make sure that the arxivId is truthy, this would be better than explicitly checking that it isn't `undefined`.

codeviking replied:
Gotcha. I go either way on this one.  I tend to avoid value coercion where I can, because it can be tricky and cause unexpected side-effects.  In this case we know that it'll be `undefined` or a non-empty string, so I like the explicitness of checking for the case that I know we don't want.

That said by the same argument I could say that there's no risk in the value coercion -- so let's go with `!!` because it's shorter (and I imagine more in line with S2's JavaScript styles!).

rreas commented on api/src/db-connection.ts:
```diff
@@ -82,13 +82,15 @@ export class Connection {
-                 WHERE symbol.paper_id = paper.s2_id
+                  FROM entity
+                 WHERE entity.paper_id = paper.s2_id
+                   AND type = 'symbol'
```
I think these queries should take into account the 'version' table. The numbers may be inflated as a single paper may have many versions with 1000s or 10000s of entities. Specifically there is an `entity.version` column and the version table itself is indexed on paper_id. Easiest thing for now is probably aggregating to max(version.version) to find the latest. Happy to address this as a follow-up.

codeviking replied:
Sure, I can take care of that before merging.

codeviking replied:
I took a stab at it -- but now that I'm reviewing it I wonder if it can even simpler.

I think y'all might be saying that there's only one `version` that I need to use (which is the most recent) for all entity types. So I can use one query to find that, and filter the counts based on that.

Let me know if that's the case I can tweak this to reflect that.

codeviking replied:
Ok, this got another tweak, I think it's right now.

rreas replied:
Yes, that looks right to me. My comment above wasn't quite right as the "version" column referenced in "entity" is actually "index" in the version table. What you've done here sidesteps the need to know that 👍 

I verified against the public schema that there are entities for papers I expect to have them (i.e., not all are 0).

codeviking replied:
I took _another_ stab at the query, with your help (it's more or less what you shared with me in Slack).

TAL.

rreas commented:
Big help, thanks!

andrewhead reviewed (commented):
Looks good! I echo Russell's comment regarding versions. Do we want to incorporate that into this PR, or a later one?

codeviking commented:
> Looks good! I echo Russell's comment regarding versions. Do we want to incorporate that into this PR, or a later one?

I'll take care of it before merging.

codeviking commented:
@andrewhead @rreas Ok, I think I incorporated the necessary changes. Please TAL.

Here's before:

![Screen Shot 2020-12-10 at 6 01 36 PM](https://user-images.githubusercontent.com/801451/101852524-b231f280-3b12-11eb-9b24-545758f71d06.png)

...and after:

![Screen Shot 2020-12-10 at 6 01 19 PM](https://user-images.githubusercontent.com/801451/101852502-a47c6d00-3b12-11eb-8003-fc5612789df9.png)

The numbers look a lot more reasonable, though 3k symbols is still a lot!

codeviking commented:
Ok, take 3.

I think I have the query correct now...and it looks like there's only one paper with symbols and citations of the latest version.

To prevent this paper from feeling less lonely I removed the bit that filters for papers who have more than 1 thing. Sure, we don't really want to show those, but at least we know what papers *have* been processed. We show the ones with data first -- so I don't think it's really too big of a deal to show the ones that don't have anything.

Here's how it looks:
![Screen Shot 2020-12-10 at 8 21 04 PM](https://user-images.githubusercontent.com/801451/101861590-3c835200-3b25-11eb-8f1c-7c34a9d4ac56.png)

Please TAL and the query and lmk if this is right.

codeviking commented:
Ok, I mucked around with the query a bit more, thanks to help from @rreas. I also:

- Added pagination
- Added counts for each entity type
- Sorting by the total # of all entities
- Added a cache-busting mechanism for the UI cache (I debated removing the cache entirely, but decided not to).

Here's how it looks:

![Screen Shot 2020-12-11 at 1 47 04 PM](https://user-images.githubusercontent.com/801451/101958112-8288f600-3bb7-11eb-86b1-da3672e2db24.png)

andrewhead commented on api/src/db-connection.ts:
```diff
@@ -77,30 +78,61 @@ export class Connection {
+             SUM(CASE WHEN entity.type = 'sentence' THEN 1 ELSE 0 END) AS sentence_count,
+             SUM(CASE WHEN entity.type = 'term' THEN 1 ELSE 0 END) AS term_count,
+             SUM(CASE WHEN entity.type = 'equation' THEN 1 ELSE 0 END) AS equation_count,
+             SUM(CASE WHEN entity.type = 'definition' THEN 1 ELSE 0 END) AS definition_count,
```
fyi: the way definitions are stored in the database is about to change with my next pipeline update. They will be sorted as rows in the `entitydata` table, where each row is with a 'term' or 'symbol' entity from the `entity` table.

Maybe it's worth removing this column for now? Or, if we want to include it anyway, we can put together a new query to get definition counts once I've incorporated the pipeline updates.

codeviking replied:
Sure. I have some work on it's way that'll fit nicely with this. I'll remove the column in that PR.

andrewhead replied:
sounds good, thank you!

andrewhead commented on ui/src/papers/index.tsx:
```diff
@@ -69,6 +73,12 @@ interface CachedPaperWithMeta {
+function cacheKey(paperId: string) {
+    // We append a version prefix so that we can bust the cache in clients when we release
+    // updates.
+    return `v0/${paperId}`; 
```
Just to check my understanding, does this mean when we release updates, we would bump the version number for the full API (say to `v0.1`) in order to bust the cache?

codeviking replied:
Yup, exactly. It's only necessary where there's a backwards incompatible change -- say, if the format of the data changes (like it did in this PR).

andrewhead commented on ui/src/papers/index.tsx:
```diff
@@ -156,65 +216,92 @@ const PaperList = () => {
+                      </React.Fragment>
+                    ))}
+                    {paper.venue && ` • ${paper.venue}`}
+                    {" •"} {paper.year}
```
Is the paper.year guaranteed to not be null in the S2 API?

codeviking replied:
I'm not sure. I don't think I changed this code, rather I just moved it. If It's `undefined` we'll just end up with a stray bullet -- a little unideal, but I *think* this page is still for internal use, so it's probably ok?  Though to your point -- when this gets baked into the UI we'll need to handle those edge cases carefully!

rreas replied:
For now, since we are focused on ArXiv papers, this value will always be available unless something has gone very, very wrong.

andrewhead replied:
Good point! Okay, I'm convinced this is okay to keep as is.

andrewhead reviewed (commented):
Thanks for this, Sam! This looks good. I added a few questions about particulars in the code.
